cpp/include/DataStorm/DataStorm.h

```diff
@@ -0,0 +1,2121 @@
+    }
+
+}
+
+//
+// Public template based API implementation
+//
+
+namespace DataStorm
+{
+
+    //
+    // Sample template implementation
+    //
+    template<typename Key, typename Value, typename UpdateTag>
+    SampleEvent Sample<Key, Value, UpdateTag>::getEvent() const noexcept
+    {
+        return _impl->event;
+    }
+
+    template<typename Key, typename Value, typename UpdateTag>
+    const Key& Sample<Key, Value, UpdateTag>::getKey() const noexcept
+    {
+        return _impl->getKey();
+    }
+
+    template<typename Key, typename Value, typename UpdateTag>
+    const Value& Sample<Key, Value, UpdateTag>::getValue() const noexcept
+    {
+        return _impl->getValue();
+    }
+
+    template<typename Key, typename Value, typename UpdateTag>
+    UpdateTag Sample<Key, Value, UpdateTag>::getUpdateTag() const noexcept
+    {
+        return _impl->getTag();
+    }
+
+    template<typename Key, typename Value, typename UpdateTag>
+    std::chrono::time_point<std::chrono::system_clock> Sample<Key, Value, UpdateTag>::getTimeStamp() const noexcept
+    {
+        return _impl->timestamp;
+    }
+
+    template<typename Key, typename Value, typename UpdateTag>
+    std::string Sample<Key, Value, UpdateTag>::getOrigin() const noexcept
+    {
+        return _impl->origin;
+    }
+
+    template<typename Key, typename Value, typename UpdateTag>
+    std::string Sample<Key, Value, UpdateTag>::getSession() const noexcept
+    {
+        return _impl->session;
+    }
+
+    template<typename Key, typename Value, typename UpdateTag>
+    Sample<Key, Value, UpdateTag>::Sample(const std::shared_ptr<DataStormI::Sample>& impl) noexcept
+        : _impl(std::static_pointer_cast<DataStormI::SampleT<Key, Value, UpdateTag>>(impl))
+    {
+    }
+
+    //
+    // Reader template implementation
+    //
+    template<typename Key, typename Value, typename UpdateTag>
+    Reader<Key, Value, UpdateTag>::Reader(Reader<Key, Value, UpdateTag>&& reader) noexcept
+        : _impl(std::move(reader._impl))
+    {
+    }
+
+    template<typename Key, typename Value, typename UpdateTag> Reader<Key, Value, UpdateTag>::~Reader()
+    {
+        if (_impl)
+        {
+            _impl->destroy();
+        }
+    }
+
+    template<typename Key, typename Value, typename UpdateTag>
+    Reader<Key, Value, UpdateTag>& Reader<Key, Value, UpdateTag>::operator=(Reader&& reader) noexcept
+    {
+        if (_impl)
+        {
+            _impl->destroy();
+        }
+        _impl = std::move(reader._impl);
+        return *this;
+    }
+
+    template<typename Key, typename Value, typename UpdateTag>
+    bool Reader<Key, Value, UpdateTag>::hasWriters() const noexcept
+    {
+        return _impl->hasWriters();
+    }
+
+    template<typename Key, typename Value, typename UpdateTag>
+    void Reader<Key, Value, UpdateTag>::waitForWriters(unsigned int count) const
+    {
+        _impl->waitForWriters(static_cast<int>(count));
```

Comment:
This static_cast is new, I have to add it because the signed/unsigned mismatch. There are a few others for other waitForXxx methods